app/src/main/java/com/android/shelfLife/model/user/UserRepositoryFirestore.kt

```diff
@@ -0,0 +1,185 @@
+package com.android.shelfLife.model.user
+
+import android.util.Log
+import com.google.firebase.auth.FirebaseAuth
+import com.google.firebase.firestore.FieldValue
+import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.firestore.ListenerRegistration
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.tasks.await
+
+class UserRepositoryFirestore(private val db: FirebaseFirestore) : UserRepository {
+
+  private val auth = FirebaseAuth.getInstance()
+  private val userCollection = db.collection("users")
+
+  // Local variable to store user data
+  private val _user = MutableStateFlow<User?>(null)
+  override val user: StateFlow<User?> = _user.asStateFlow()
+
+  // Invitations StateFlow
+  private val _invitations = MutableStateFlow<List<String>>(emptyList())
+  override val invitations: StateFlow<List<String>> = _invitations.asStateFlow()
+
+  // Listener for invitations
+  private var invitationsListenerRegistration: ListenerRegistration? = null
+
+  override fun getNewUid(): String {
+    return userCollection.document().id
+  }
+
+  override suspend fun initializeUserData() {
+    val currentUser = auth.currentUser ?: throw Exception("User not logged in")
+    try {
+      // Fetch user data from Firestore
+      val snapshot = userCollection.document(currentUser.uid).get().await()
+      if (snapshot.exists()) {
+        val userData = snapshot.toObject(User::class.java)
+        _user.value = userData
+        _invitations.value = userData?.invitationUIDs ?: emptyList()
+      } else {
+        _user.value = null
+        _invitations.value = emptyList()
+      }
+    } catch (e: Exception) {
+      Log.e("UserRepository", "Error initializing user data", e)
+      _user.value = null
+      _invitations.value = emptyList()
+      throw e
+    }
+  }
+
+  override fun startListeningForInvitations() {
+    val currentUser = auth.currentUser
+    if (currentUser != null) {
+      // Remove any existing listener
+      invitationsListenerRegistration?.remove()
```

Comment:
nit picky: maybe it's better to call `stopListeningForInvitations()` here? Maybe there is a reason you explicitly wrote the removal line again